---
title: "Naming convention"
format:
  html:
    code-fold: true
    code-summary: "Show the code"
    embed-resources: true
editor: visual
vignette: >
  %\VignetteIndexEntry{audit}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{quarto::html}
---

```{r}
library(kintsugi)
```

`kintsugi` uses the following naming conventions.

## Data sources

-   **dnadb:** data from the DNA Database, also referred to as DLIMS (DNA Laboratory Information Management System)

-   **igene:** data from the iGene database

-   **csv:** data from csv files

## Data tables

Data tables accessed from the DNA Database are named the same in `kintsugi` as their original names, only, whilst DNA Database tables are named in CamelCase, tables in `kintsugi` follow the [tidyverse style guide](https://style.tidyverse.org/syntax.html) by using snake_case.

```{r}
#| include: TRUE

dnadb_results_access <- connect_to_lazy_tbl("ResultsAccess")

```

## Identifiers

-   **nhsno:** the patient's 10 digit NHS number.

-   **rno:** the iGene referral (R) number.
    This number identifies a referral - one referral can have multiple samples.

-   **sno:** the iGene sample (S) number.
    This number identifies a sample, such as shavings from a formalin-fixed paraffin-embedded (FFPE) block.
    One sample can have multiple derivatives.

-   **dno:** the iGene derivative (D) number.
    This number identifies material that is *derived* from a sample, such as DNA which is extracted from FFPE shavings.

-   **tno:** the iGene test (T) number.
    This number identifies a single test instance.
    One D number can have multiple T numbers.

-   **labno:** the laboratory number or "DNA number" used on DNA Database.
    Any labno is an 8-digit string of numbers.
    The first two digits indicate the year the sample was received.
    The laboratory number is the DNA Database equivalent of the iGene D number.

-   **pathno:** the identifier of the FFPE pathology block from which DNA was extracted.
    Different pathology labs have different formats for pathno.

-   **ws:** the worksheet the sample was tested on at the Manchester Genomic Laboratory Hub.
    The format is WS followed by 6 digits.

## Verbs

I have tried to use consistent verbs for dataframes and functions.
These verbs are based on their usage in tidyverse packages such as `dplyr` and `tidyr`.

-   **bind:** bind together two (or more) large datasets, using the `rbind` or `purrr:list_rbind()` function.

-   **clean:** filter a table, removing duplicates or tidying up free-text columns

-   **collect:** pull information out of a database (`dplyr::collect`).

-   **connect:** establish a connection with a database table but without collecting data (a "lazy table") using `dplyr::tbl`

-   **join:** bring two tables together using a common key variable (`dplyr::left_join`)

-   **list:** list filepaths using `base::list.files`

-   **pivot:** restrcucture a table using `tidyr::pivot_longer` or `tidyr::pivot_wider`

-   **mutate:** add on useful information to a dataframe (`dplyr::mutate`).

-   **read:** reading information from a single file (`dplyr::read_csv`)

## Projects

-   **service:** samples tested as part of the live clinical service, with a genetic result returned to a clinician.

-   **validation:** samples tested as part of a validation of a method, with no genetic result returned to a clinician.

## Acronyms

-   **orpp:** One Result Per Patient

-   **orpr:** One Result Per Referral

-   **orps:** One Result Per Sample

## Ovarian cancer terms

-   **glvar:** germ-line variant data.
    Data generated by next generation sequencing for analysis of germline (i.e inherited) sequence variants.
    In ovarian cancer testing, this refers to test directory indication R207 for inherited ovarian cancer.

-   **tvar:** tumour variant data.
    Data generated by next generation sequencing of DNA from formalin-fixed paraffin-embedded (FFPE) tumour tissue for analysis of somatic sequence variants.
    In ovarian cancer testing, this may be referred to as "tumour BRCA" or "tBRCA", but I have chosen "tvar" as variants in non-BRCA genes may also be detected.

-   **gi:** genomic instability results.
    Results for assessment of genomic instability (GI), such as using shallow whole genome sequencing with the SeqOne SomaHRD pipeline.

-   **ICP:** Inherited Cancer Panel

# File naming

The naming convention for files and tables is: genetic test, then data source, then what has been done to the data.

Example: "tvar_dnadb_cleaned" - tumour variant data sourced from DNA Database which has been cleaned.
